File: src/main/java/com/roscopeco/jasm/ThrowingErrorListener.java
/*
 * Copyright (c)2022 Ross Bamford & Contributors
 *
 * Licensed under the MIT license. See LICENSE.md for details.
 */
package com.roscopeco.jasm;

import lombok.NonNull;
import org.antlr.v4.runtime.BaseErrorListener;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.Recognizer;

class ThrowingErrorListener extends BaseErrorListener {
    private final String unitName;

    ThrowingErrorListener(@NonNull final String unitName) {
        this.unitName = unitName;
    }

    @Override
    public void syntaxError(
            final Recognizer<?, ?> recognizer,
            final Object offendingSymbol,
            final int line,
            final int charPositionInLine,
            final String msg,
            final RecognitionException e
    ) {
        throw new SyntaxErrorException(unitName + ": line " + line + ":" + charPositionInLine + " " + msg, e);
    }
}

File: src/main/kotlin/com/roscopeco/jasm/ThrowingErrorListener.kt
/*
 * Copyright (c)2022 Ross Bamford & Contributors
 *
 * Licensed under the MIT license. See LICENSE.md for details.
 */
package com.roscopeco.jasm

import org.antlr.v4.runtime.BaseErrorListener
import org.antlr.v4.runtime.RecognitionException
import org.antlr.v4.runtime.Recognizer

internal open class ThrowingErrorListener(private val unitName: String) : BaseErrorListener() {
    override fun syntaxError(
        recognizer: Recognizer<*, *>?,
        offendingSymbol: Any,
        line: Int,
        charPositionInLine: Int,
        msg: String,
        e: RecognitionException?
    ) = throw SyntaxErrorException("$unitName: line $line:$charPositionInLine $msg", e ?: RuntimeException())
}